apps/common/services.py

```diff
@@ -0,0 +1,111 @@
+
+    async def create_many(self, *, session: AsyncSession, objs: List[CreateSchemaType]) -> Union[List[ModelType], None]:
+        insert_statement = (
+            insert(self.model)
+                .values([jsonable_encoder(obj=obj, exclude_unset=True, by_alias=False) for obj in objs])
+                .returning(self.model)
+        )
+        statement = select(self.model).from_statement(insert_statement).execution_options(populate_existing=True)
+        result: ChunkedIteratorResult = await session.execute(statement=statement)
+        await session.commit()
+        data: Union[List[ModelType], None] = result.scalars().all()
+        return data
+
+    async def read(self, *, session: AsyncSession, data: Union[Dict, None] = None,
+                   obj: Union[ReadSchemaType, None] = None) -> Union[ModelType, None]:
+        data = data if data else {}
+        obj_in_data = jsonable_encoder(obj=obj, exclude_unset=True, by_alias=False) if obj else {}
+        where_dict = {**data, **obj_in_data}
+        statement = select(self.model)
+        model = self.model.c if isinstance(self.model, Table) else self.model
+        for k, v in where_dict.items():
+            statement = statement.where(getattr(model, k) == v)
+        result: ChunkedIteratorResult = await session.execute(statement=statement)
+        response: Union[ModelType, None] = result.first()
+        return response if isinstance(self.model, Table) else response[0]
+
+    async def update(self, *, session: AsyncSession, data: Union[Dict, None] = None,
+                     obj: Union[Dict, UpdateSchemaType] = None) -> Union[ModelType, None]:
+        obj = obj if obj else {}
+        data = data if data else {}
+        values = obj if isinstance(obj, Dict) else jsonable_encoder(obj=obj, exclude_unset=True, by_alias=False)
+        where_expr = []
+        for k, v in data.items():
+            where_expr.append(getattr(self.model, k) == v)
+        update_statement = (
+            update(self.model)
+                .where(and_(*where_expr))
+                .values(**values)
+                .returning(self.model)
+                .execution_options(synchronize_session="fetch")
+        )
+        statement = (
+            select(self.model).from_statement(statement=update_statement).execution_options(populate_existing=True)
+        )
+        result: ChunkedIteratorResult = await session.execute(statement=statement)
+        await session.commit()
+        data: Union[ModelType, None] = result.scalar_one_or_none()
+        return data
+
+    async def delete(self, *, session: AsyncSession, data: Union[Dict, None] = None,
+                     obj: Union[DeleteSchemaType, None] = None) -> CursorResult:
+        data = data if data else {}
+        schema = jsonable_encoder(obj=obj, exclude_unset=True, by_alias=False) if obj else {}
+        where_dict = {**data, **schema}
+        statement = delete(self.model)
+        for k, v in where_dict.items():
```

Comment:
You can try to refactor this logic, DRY pattern.

Reply:
ok